src/uct/self/self_ep.c

```diff
@@ -23,19 +23,81 @@ static UCS_CLASS_CLEANUP_FUNC(uct_self_ep_t)
     ucs_trace_func("self=%p", self);
 }
 
-UCS_CLASS_DEFINE(uct_self_ep_t, uct_base_ep_t)
+UCS_CLASS_DEFINE(uct_self_ep_t, uct_base_ep_t);
 UCS_CLASS_DEFINE_NEW_FUNC(uct_self_ep_t, uct_ep_t, uct_iface_t *,
                           const uct_device_addr_t *, const uct_iface_addr_t *);
 UCS_CLASS_DEFINE_DELETE_FUNC(uct_self_ep_t, uct_ep_t);
 
+static void
+uct_self_ep_am_reserve_and_get_new_buffer(ucs_status_t *status,
+                                          uct_self_iface_t *self_iface,
+                                          const uct_self_msg_desc_t *desc)
+{
+    /* Reserve the buffer */
+    uct_recv_desc_iface(desc->am_recv_data) = &self_iface->super.super;
+
+    /* get new buffer from memory pool
+    * No need to copy data from payload to desc->am_recv_data */
+    self_iface->msg_cur_desc = ucs_mpool_get(&self_iface->msg_desc_mp);
+    VALGRIND_MAKE_MEM_DEFINED(self_iface->msg_cur_desc,
+                              sizeof(*(self_iface->msg_cur_desc)));
+    if (NULL == self_iface->msg_cur_desc) {
+        ucs_error("Failed to get next descriptor in SELF MP storage");
+        *status = UCS_ERR_NO_RESOURCE;
```

Comment:
if ser callback returns INPROGRESS it means user took ownership of the buffer, but it you fail to allocate a new buffer to replace it, you would return NO_RESOURCE. But the data was delivered to the user callback already (and the user can process it at any time and call release_desc). The user would call send_am again because it returned NO_RESOURCE, and it will result in double send.